cf/data/utils.py

```diff
@@ -10,6 +10,56 @@
 from ..units import Units
 
 
+def _is_numeric_dtype(array):
+    """True if the given array is of a numeric dtype.
+
+    .. versionadded:: 4.0.0
+
+        :Parameters:
+
+            array: numpy-like array
+
+        :Returns:
+
+            `bool`
+                Whether or not the array holds numeric elements.
+
+    **Examples:**
+
+    >>> a = np.array([0, 1, 2])
+    >>> _is_numeric_dtype(a)
+    True
+    >>> a = np.array([False, True, True])
+    >>> _is_numeric_dtype(a)
+    True
+    >>> a = np.array(["a", "b", "c"], dtype="S1")
+    >>> _is_numeric_dtype(a)
+    False
+    >>> a = np.ma.array([10.0, 2.0, 3.0], mask=[1, 0, 0])
+    >>> _is_numeric_dtype(a)
+    True
+    >>> a = np.array(10)
+    >>> _is_numeric_dtype(a)
+    True
+    >>> a = np.empty(1, dtype=object)
+    >>> _is_numeric_dtype(a)
+    False
+
+    """
+    # TODODASK: do we need to make any specific checks relating to ways of
+    # encoding datetimes, which could be encoded as strings, e.g. as in
+    # "2000-12-3 12:00", yet could be considered, or encoded as, numeric?
+    dtype = array.dtype
+    # This checks if the dtype is either a standard "numeric" type (i.e.
+    # int types, floating point types or complex floating point types)
+    # or Boolean, which are effectively a restricted int type (0 or 1).
+    # We determine the former by seeing if it sits under the 'np.number'
+    # top-level dtype in the NumPy dtype hierarchy; see the
+    # 'Hierarchy of type objects' figure diagram under:
+    # https://numpy.org/doc/stable/reference/arrays.scalars.html#scalars
+    return np.issubdtype(dtype, np.number) or np.issubdtype(dtype, np.bool_)
```

Comment:
I tried to cover that further into the comment above, "or Boolean, which are effectively a restricted int type (0 or 1)." Zeroes and ones are numeric to me, fundamentally, and indeed I noticed when checking for any similar functions existing in Dask or NumPy that (they don't seem to define their own, but) Dask imports the pandas function `is_numeric_dtype` (same name I have used, but internally marked) and they [also consider Boolean as numeric](https://github.com/pandas-dev/pandas/blob/945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5/pandas/core/dtypes/common.py#L1188-L1227 ).

Happy to remove it though if you have a decent reason that Booleans shouldn't be considered numeric? Or could add a flag kwarg to toggle such behaviour?

Reply:
Hmm, more confusion, but not of your making! cfdm and cf short circuit on non-equal data types: https://github.com/NCAS-CMS/cfdm/blob/master/cfdm/data/data.py#L2256

I recall that that was deliberate, as cf equality goes beyond just numerical equivalence. For example, the units have to be equal and not just equivalent. Image that we have data from two netCDF files that have different precisions but equivalent values, are these variables equal? We thought not, since precision can sometimes matter beyond just noisy bits when you actually do stuff with your data - e.g. the global integral of a 64-bit field would not be the same as the same field truncated to 32-bits, even though numpy might say that the fields are equal element-wise.

What do you think of that history?

Your suggestion of new keywords to control this (NaNa, infs, dtypes, and even units) are sounding like good ideas to me right now ... but what would the defaults be?